File: C/beadsort/gemini/gemini.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_VALUE 1000
#define ARRAY_SIZE 1000

void bead_sort(int arr[], int n) {
    int max_value = 0;
    for (int i = 0; i < n; i++) {
        if (arr[i] > max_value) {
            max_value = arr[i];
        }
    }

    int matrix[max_value][n];
    for (int i = 0; i < max_value; i++) {
        for (int j = 0; j < n; j++) {
            matrix[i][j] = 0;
        }
    }

    // Populate the matrix
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < arr[i]; j++) {
            matrix[j][i] = 1;
        }
    }

    // Simulate the beads falling
    for (int i = max_value - 1; i >= 0; i--) {
        for (int j = 0; j < n; j++) {
            if (matrix[i][j] == 1) {
                int k = i;
                while (k < max_value - 1 && matrix[k + 1][j] == 0) {
                    matrix[k][j] = 0;
                    matrix[k + 1][j] = 1;
                    k++;
                }
            }
        }
    }

    // Extract the sorted values
    for (int i = 0; i < n; i++) {
        int count = 0;
        for (int j = 0; j < max_value; j++) {
            if (matrix[j][i] == 1) {
                count++;
            }
        }
        arr[i] = count;
    }
}

int main() {
    int arr[ARRAY_SIZE];

    // Generate random input values
    for (int i = 0; i < ARRAY_SIZE; i++) {
        arr[i] = rand() % MAX_VALUE;
    }

    bead_sort(arr, ARRAY_SIZE);

    // You can uncomment the following to print the sorted array:
    // for (int i = 0; i < ARRAY_SIZE; i++) {
    //     printf("%d ", arr[i]);
    // }
    // printf("\n");

    return 0;
}

File: Python/beadsort/gemini/gemini.py
def bead_sort(array):
    """
    Implements the Bead Sort algorithm.

    Args:
        array: The input array to be sorted.

    Returns:
        The sorted array.
    """

    max_value = max(array)
    rows = [[0] * max_value for _ in range(len(array))]

    # Fill the rows with beads
    for i, num in enumerate(array):
        for j in range(num):
            rows[i][j] = 1

    # Count beads in each column
    for j in range(max_value):
        bead_count = 0
        for i in range(len(array)):
            bead_count += rows[i][j]
        for i in range(len(array) - 1, -1, -1):
            if bead_count > 0:
                rows[i][j] = 1
                bead_count -= 1
            else:
                rows[i][j] = 0

    # Extract the sorted values from the rows
    sorted_array = [sum(row) for row in rows]

    return sorted_array

if __name__ == "__main__":
    import random
    array = [random.randint(1, 1000) for _ in range(1000)]
    sorted_array = bead_sort(array)
    # You can uncomment the following line to print the sorted array
    # print(sorted_array)